ADR034: Account Rekeying

sunnya97 opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

## Description

This PR includes an ADR for a new MsgType to allow the updating of the public key for an account.

Corresponding implementation in #7452

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

clevinson commented:
Thanks @sunnya97!

Can you update the # of this to ADR034 ? We currently have WIP ADR's up to ADR033 already as active pull requests.

robert-zaremba commented on docs/architecture/adr-032-change-pubkey.md:
```diff
@@ -0,0 +1,81 @@
+
+Breaks the current assumed relationship between address and pubkeys as H(pubkey) = address. This has a couple of consequences.
+
+* We cannot prune accounts with 0 balance that have had their pubkey changed (we currently do not currently do this anyways, but the reason we have account numbers is presumably for this purpose).
```
We can still make it possible to prune accounts by introducing a new message: `MsgNukeAccount`. When an author will delete HIS account, then the deposit is released.

sunnya97 replied:
I implemented the "deposit" as an extra Gas charge rather than an actual monetary deposit.  With a gas charge, we could provide a gasrefund when an account is nuked, but that could potentially turn into GasToken shenanigans like on Ethereum.

robert-zaremba commented:
I really like this ADR. 

I also started drafting a similar thing 2 weeks ago: to decouple account address from a public key and proposed it as one of the goals for SDK. I call it re-keying (this is in fact a term used by Algorand). During the review of #7086 I noted we need to support accounts with native addresses, specifically:

* multisigs with _native_ public key
*  accounts which will enable dynamic keys.

@aaronc  in https://github.com/cosmos/cosmos-sdk/pull/7086#issuecomment-705724579 pointed to a Group Module we are designing at Regen Network, which solves the latter point  using a separation of an account object from the policies. 

What's good, all this ideas will play together very well:

* we can use the `Address` algorithm from ADR-028 to compute an initial address.
* we can update all `account` structures to add a fixed `address` field, which will be computed using the aforementioned `Address` function.
* All accounts will support re-keying.

aaronc commented on docs/architecture/adr-034-change-pubkey.md:
````diff
@@ -0,0 +1,81 @@
+
+As an example, account pubkey change message can be defined as follows.
+
+```json
+{
+    "type": "cosmos-sdk/StdTx",
+    "value": {
+        "address": "cosmos1wf5h7meplxu3sc6rk2agavkdsmlsen7rgsasxk",
+        "public_key": "cosmospub1addwnpepqdszcr95mrqqs8lw099aa9h8h906zmet22pmwe9vquzcgvnm93eqygufdlv"
+    },
+    "signature": "a9n7pIqCUuYJTCm7ZBv1cqqlM3uYyX/7SnaSXA8zrG0CBWP6p55pTFFHYn39tVvFtRbGE7gXF1qCiaOilJ8NtQ=="
+}
+```
````
This isn't really a valid transaction. Let's just document that `MsgChangePubKey` transaction needs to be sign by the existing pubkey in state.
```suggestion
```

aaronc commented on docs/architecture/adr-034-change-pubkey.md:
```diff
@@ -0,0 +1,81 @@
+
+## Decision
+
+We propose the creation of a new feature called `changepubkey` that is an extension to `auth` that allows accounts to update the public key associated with their account, while keeping the address the same.
```
```suggestion
We propose the creation of a new feature  to `x/auth` that allows accounts to update the public key associated with their account, while keeping the address the same.
```

aaronc reviewed (changes requested):
Concept ACK

Can you add a boolean `EnableChangePubKey` param?

aaronc commented on docs/architecture/adr-034-change-pubkey.md:
```diff
@@ -0,0 +1,63 @@
+
+## Decision
+
+We propose the creation of a new feature  to `x/auth` that allows accounts to update the public key associated with their account, while keeping the address the same.  This feature can be enabled using an `EnableChangePubKey` param.
```
```suggestion
We propose the addition of a new feature to `x/auth` that allows accounts to update the public key associated with their account, while keeping the address the same.  This feature can be enabled using an `EnableChangePubKey` param.
```

aaronc commented on docs/architecture/adr-034-change-pubkey.md:
```diff
@@ -0,0 +1,63 @@
+
+Once, approved, the handler for this message type, which takes in the AccountKeeper, will update the in-state pubkey for the account and replace it with the pubkey from the Msg.
+
+Because an account can no longer be pruned from state once its pubkey has changed, we can charge an additional gas fee for this operation to compensate for this this externality (this bound gas amount is configured as parameter `PubKeyChangeCost`). The bonus gas is charged inside handler, using the `ConsumeGas` function.
```
```suggestion
Because an account can no longer be pruned from state once its pubkey has changed, we can charge an additional gas fee for this operation to compensate for this this externality (this bound gas amount is configured as parameter `PubKeyChangeCost`). The bonus gas is charged inside the handler, using the `ConsumeGas` function.
```

aaronc reviewed (approved):
ACK

One question: do we allow two accounts with the same pubkey but different addresses?

zmanian commented:
I'm mostly comfortable with this ADR.

here is my reasoning.

1. It's presumably not a mandatory for a chain to implement this message type.

2. I'd oppose implementing it on the hub without other fee improvements like consensus min fee but all of those things are coming.

sunnya97 commented:
@aaronc Yeah, I think that should be allowed.  I can't think of a reason it shouldn't be, as msgs explicitly declare the originating address, and don't depend on the signature/pubkey for that

sunnya97 commented on docs/architecture/adr-034-change-pubkey.md:
```diff
@@ -0,0 +1,63 @@
+
+## Decision
+
+We propose the addition of a new feature to `x/auth` that allows accounts to update the public key associated with their account, while keeping the address the same.  This feature can be enabled using an `EnableChangePubKey` param.
```
https://twitter.com/sunnya97/status/1112464889300291586

zmanian commented:
One of the big utilities I see for this is it makes solo machines much more useful.

robert-zaremba commented on docs/architecture/adr-034-change-pubkey.md:
```diff
@@ -0,0 +1,63 @@
+# ADR 034: Change PubKey
```
Please rename the title to something more specific. I propose `Account Rekeying`. Or add some context, eg: `Change PubKey Msg`, `Change PbKey  service` ...

I think Algorand originally introduced this feature and coined the term [Rekeying](https://www.algorand.com/resources/blog/announcing-rekeying)

robert-zaremba commented on docs/architecture/adr-034-change-pubkey.md:
```diff
@@ -0,0 +1,63 @@
+## Status
+
+Proposed
+
```
We require a summary section. Please have a look at the [template](https://github.com/cosmos/cosmos-sdk/blob/master/docs/architecture/adr-template.md)

robert-zaremba commented on docs/architecture/adr-034-change-pubkey.md:
```diff
@@ -0,0 +1,63 @@
+
+## Context
+
+Currently, in the Cosmos SDK, the address of an auth account is always based on the hash of the public key.  Once an account is created, the public key for the account is set in stone, and cannot be changed.  This can be a problem for users, as key rotation is a useful security practice, but is not possible currently.  Furthermore, as multisigs are a type of pubkey, once a multisig for an account ...
```
```suggestion
Currently, in the Cosmos SDK, the address of an auth `BaseAccount` is based on the hash of the public key.  Once an account is created, the public key for the account is set in stone, and cannot be changed.  This can be a problem for users, as key rotation is a useful security practice, but is not possible currently.  Furthermore, as multisigs are a type of pubkey, once a multisig for an account is set, it can not be updated.  This is problematic, as multisigs are often used by organizations or companies, who may need to change their set of multisig signers for internal reasons.
```

robert-zaremba commented on docs/architecture/adr-034-change-pubkey.md:
````diff
@@ -0,0 +1,63 @@
+To build this system, we design a new Msg type as follows:
+
+```protobuf
+message MsgChangePubKey {
+  string address = 1;
+  google.protobuf.Any pub_key = 2;
+}
````
:+1:

robert-zaremba commented on docs/architecture/adr-034-change-pubkey.md:
```diff
@@ -0,0 +1,63 @@
+
+Once, approved, the handler for this message type, which takes in the AccountKeeper, will update the in-state pubkey for the account and replace it with the pubkey from the Msg.
+
+Because an account can no longer be pruned from state once its pubkey has changed, we can charge an additional gas fee for this operation to compensate for this this externality (this bound gas amount is configured as parameter `PubKeyChangeCost`). The bonus gas is charged inside the handler, using the `ConsumeGas` function.
```
Please create a new section for the fee operations. It seams that this is not implemented and we will need to specify it correctly. How about: "Further Discussions" Section?

sunnya97 replied:
What do you mean by fee operations?  This handler doesn't deduct fees.  It just uses an extra constant gas amount, which then is converted to additional fee deducted by auth.

robert-zaremba replied:
I suggest to add a new, short section to elaborate more on the consequences that we can't prune accounts, and maybe include a bit from this comment: https://github.com/cosmos/cosmos-sdk/pull/7491#discussion_r502601036

robert-zaremba commented on docs/architecture/adr-034-change-pubkey.md:
```diff
@@ -0,0 +1,63 @@
+* While the purpose of this is intended to allow the owner of an account to update to a new pubkey they own, this could technically also be used to transfer ownership of an account to a new owner.  For example, this could be use used to sell a staked position without unbonding or an account that has vesting tokens.  However, the friction of this is very high as this would essentially have to be d ...
+* Will require that PubKeys for an account are included in the genesis exports.
+
+## References
```
```suggestion
## References

+ https://www.algorand.com/resources/blog/announcing-rekeying
```

robert-zaremba reviewed (changes requested):
Overall all good.

Requested changes:
+ update to the latest ADR template
+ Update the spec to use `Msg` service, rather then legacy handlers (we are marking them deprecated)
+ Add a section for future discussions (re: fees)

Suggestions:
+ Let's update the title

robert-zaremba commented on docs/architecture/adr-034-account-rekeying.md:
```diff
@@ -0,0 +1,75 @@
+
+Breaks the current assumed relationship between address and pubkeys as H(pubkey) = address. This has a couple of consequences.
+
+* We cannot prune accounts with 0 balance that have had their pubkey changed (we currently do not currently do this anyways, but the reason we have account numbers is presumably for this purpose).
```
```suggestion
* We cannot prune accounts with 0 balance that have had their pubkey changed. Currently, we do not prune accounts anyways, but the reason we have account numbers is presumably for this purpose.
```

robert-zaremba reviewed (changes requested):
1. Before merging we can change a status to accepted. This is the syntax for describing the status: `{CONSENSUS STATUS} {IMPLEMENTATION STATUS}`. More about it [here](https://github.com/cosmos/cosmos-sdk/blob/master/docs/architecture/PROCESS.md#adr-status).
2. Please add the entry to the index (README.md)
3. Consider adding a section for further discussion related to pruning accounts. If you think it's not needed then fair enough.

alessio commented on docs/architecture/adr-034-account-rekeying.md:
```diff
@@ -0,0 +1,75 @@
+
+### Positive
+
+* Will allow users and validator operators to employ better operational security practices with key rotation.
```
Sunny, mind elaborating on this? What's the plan wrt key rotation?

ethanfrey replied:
validator operators typically use multisigs. They could just move to new on-chain multisig as soon as they are available.

sunnya97 replied:
But if I'm already running a validator, how would I switch to an on-chain multisig when my operator address is currently an account with a traditional multisig as its pubkey?

alexanderbez replied:
We could instead devise a way to provide a means of one-time "migrating".

sunnya97 replied:
If we allow for one time migration, while keeping the same address, we've already accepted the primary negative of this feature (breaking the relationship between pubkey and address).  At that point, what's the point of preventing further rekeying?

alessio commented:
Just FTR: I think this ADR, if approved and implemented as-is, would make https://github.com/cosmos/cosmos-sdk/pull/7727's objective almost impossible to be achieved.

alexanderbez commented on docs/architecture/adr-034-account-rekeying.md:
```diff
@@ -0,0 +1,75 @@
+
+Once, approved, the handler for this message type, which takes in the AccountKeeper, will update the in-state pubkey for the account and replace it with the pubkey from the Msg.
+
+Because an account can no longer be pruned from state once its pubkey has changed, we can charge an additional gas fee for this operation to compensate for this this externality (this bound gas amount is configured as parameter `PubKeyChangeCost`). The bonus gas is charged inside the handler, using the `ConsumeGas` function.
```
`Because an account can no longer be pruned from state once its pubkey has changed...`. Can you please elaborate on this.

robert-zaremba replied:
Check this comment, and the link: https://github.com/cosmos/cosmos-sdk/pull/7491#discussion_r519160750

It's about remembering which address is used with which PK. I believe this is the scenario:
1. you have `account1 = {pk1}`
2. your `(pk1, sk1)` was compromised
3. you rekey: `account1 = {pk2}`
4. we want to assure that `pk1`  won't be used in the future. (otherwise it could do actions for `account1`, even if it was deleted).

Maybe there are cases, which will be good to elaborate.

alexanderbez replied:
Yes, I'm aware of this situation -- I just meant the ADR should describe it because not everyone has this context 👍

ethanfrey commented:
> Just FTR: I think this ADR, if approved and implemented as-is, would make #7727 objective almost impossible to be achieved.

I agree there. It is not just ADR 036, but a large swath of possible client-side improvements.

Short: There have been very limited js libs for cosmos for historical reasons. We have worked hard on CosmJS to give a good foundation for all to build more client tooling. One big hope was coordination off-chain - things like authenticating as in ADR 036, but also wallet connect, creating multisigs, etc.

I think this will severely impact many layer 2 extensions. I wonder how algorand handles that (or they have a much simpler state machine).

There is a nice separation in Ethereum from "external accounts" - which are tied 1:1 with a pubkey for authorization and "contract accounts' - which are controlled by logic. Regen's Group account would enable swapping pubkeys among other things, but would fit in the "contract accounts" concept - and never be tied to one pubkey. The same as various CosmWasm solutions providing such rotation (like cw1-whitelist).

By making "external accounts" mutable, you break some invariants of the blockchain, which at first do not seem to have problems, but may raise many subtle issues - all design was made with them in mind. I would be much more comfortable if these are flagged as "contract accounts" and not mixed with typical "external accounts"

robert-zaremba commented:
@ethanfrey - IMHO, having a clear separation of a public key and an account is highly desirable. We should, if possible, design for that separation. Ethereum is not a best example. Externally Owned Accounts is a low level primitive, which shouldn't be exposed to the users. It caused strict patterns which then caused hacks. First and foremost, many wallets assumed that user is a (pk,sk) pair. That is a very big limitation.
To list few more:
* hacks with `address.call.send`, `address.call.transfer`, `address.call.value(amount)().transfer` .... - with all this is backward engineered to assume we are sending ETH to EOA.
* EOA, by design, can't have a logic and there is no single standard for account with logic (on-chain wallet).
* if you want to go to market to an average Joe, you need to have that wallet abstraction, with some recovery... guess what happens - everyone is doing their own version of on-chain wallet.
    * now we would like to use that wallet as an identity ... lot of complexity arise, because we need to do low level calls / hack to treat that wallet as a proxy for arbitrary calls. In other words, we want to call all other ethereum contracts through that wallet.

ethanfrey commented:
Robert, I don't say Ethereum is perfect, and the points you raise have little to do with key rotation.

What I am saying is this separation of concepts has allowed people to reason about behavior - especially on the off-chain / wallet / layer 2 side. I do not see any clear analysis of how one could build off-chain infrastructure with this key rotations.

e.g. if I open a state channel with an account, I would only open it with a pubkey tied to an EOA. I cannot open it with some on-chain multisig contract as one party. So, I design a state channel solution with that understanding. Suddenly, even these stable addresses may switch out pubkeys while packets are in transit in Layer 2. How can I build a state channel that is secure in that situation?

I think this ADR should have detailed analysis and proposed solutions for a few such problems. Basically any layer 2 solution will face these issue.

ethanfrey commented on docs/architecture/adr-034-account-rekeying.md:
```diff
@@ -0,0 +1,75 @@
+
+## Context
+
+Currently, in the Cosmos SDK, the address of an auth `BaseAccount` is based on the hash of the public key.  Once an account is created, the public key for the account is set in stone, and cannot be changed.  This can be a problem for users, as key rotation is a useful security practice, but is not possible currently.  Furthermore, as multisigs are a type of pubkey, once a multisig for an account  ...
```
Flexible on chain multisigs are done in CosmWasm and coming soon in SDK. I don't see this point as a reason to add the pubkey rotation which has many negative impacts on client code.

jackzampolin replied:
Tend to agree here @ethanfrey

ethanfrey commented on docs/architecture/adr-034-account-rekeying.md:
```diff
@@ -0,0 +1,75 @@
+Breaks the current assumed relationship between address and pubkeys as H(pubkey) = address. This has a couple of consequences.
+
+* We cannot prune accounts with 0 balance that have had their pubkey changed (we currently do not currently do this anyways, but the reason we have account numbers is presumably for this purpose).
+* This makes wallets that support this feature more complicated. For example, if an address on chain was updated, the corresponding key in the CLI wallet also needs to be updated.
```
It is far more than a UI issue. If you have any intelligent client node, like a long-living program, especially if it does any caching or rollup. This will make life harder, possibly impossible to resolve.

ethanfrey reviewed (commented):
I think this can cause lots of unintended and undiscovered impacts.

Please take a look at https://github.com/CosmWasm/cosmwasm-plus/tree/master/contracts/cw1-whitelist which is a super minimal contract. If you give tokens to this, it provides one clear address and you can swap under that. However, the account address *never* has a pubkey associated with it. And it actually provides more options that what you describe here and is easily extended in eg [`cw1-subkeys`](https://github.com/CosmWasm/cosmwasm-plus/tree/master/contracts/cw1-subkeys)

I think you could do something very similar in a native module, making a lot of "sub-module addresses", which may have a rotating pubkey (or keys) behind them. But since the address *never* has a pubkey directly when querying `x/auth`, then you never create confusion. Either an account is controlled by an off-chain pubkey or controlled by on-chain logic.

jackzampolin commented:
I tend to agree with @ethanfrey here, this is pretty huge change and it seems like the usecases we are envisioning are taken care of by the group module. Breaking a bunch of client side assumptions right as we are getting to stable client side libs seems bad.

robert-zaremba commented:
During the call I mentioned that I made a comment with few ideas how to support verifiable off-chain signatures:
1. Attach blockchain transaction ID to the `MsgSignData` (as a required field). Verifier checks with the blockchain that the transaction exists and is signed by the same key.
1. Similarly to the above one: instead of attaching a transaction ID, we can attach a merkle proof + block height, proving that at that time, an account has assigned correct public key.
1. Update this ADR (ADR-34) to store the past public keys. If someone will rekey, we will have a stack of the previous keys. Dapp can query that stack to search for a public key.

I think the first one is the easiest one. Note it's also sound, because whenever a user will rekey, he will need to do a transaction with his old key. So for every users' public key, if the key was associated with a blockchain account, there will be always an associated blockchain transaction.

aaronc commented on docs/architecture/adr-034-account-rekeying.md:
```diff
@@ -0,0 +1,75 @@
+
+## Decision
+
+We propose the addition of a new feature to `x/auth` that allows accounts to update the public key associated with their account, while keeping the address the same.  This feature can be enabled using an `EnableChangePubKey` param.
```
In the call it seemed like people thought this `EnableChangePubKey` param was superfluous... but I'm also not sure it hurts to have it.

sunnya97 commented:
@ethanfrey 

> There is a nice separation in Ethereum from "external accounts" - which are tied 1:1 with a pubkey for authorization and "contract accounts' - which are controlled by logic.

There is a push in the Ethereum to actually remove EOAs and have only contract accounts (see [EIP 2938 - Account Abstraction](https://eips.ethereum.org/EIPS/eip-2938))

>  Client tooling. One big hope was coordination off-chain - things like authenticating as in ADR 036, but also wallet connect, creating multisigs, etc.

By designing client tooling to be focused on working with accounts that cannot change pubkeys, this will likely leave accounts that can (on-chain multisigs, smart contract wallets, etc) out in the dust.  This is not desirable (this happens already on some off-chain coordination tools like Snapshot).  Instead, we should make it clear that it is expected that in the general-case, the authentication logic of accounts is mutable and composed of logic.   And for applications where this does not work, these applications should explicitly special case fixed pubkeys in their modules.  See next section:

> especially on the off-chain / wallet / layer 2 side. I do not see any clear analysis of how one could build off-chain infrastructure with this key rotations. ... e.g. if I open a state channel with an account, I would only open it with a pubkey tied to an EOA.

For things that need long term stable keys, their modules should explicitly lock users into such keys.  For example, when designing a payment channel module, the payment channel shouldn't be based on the public key of the accounts, but rather a new public key should be stored in the Payment channel struct.  This could be the same public key as the account's public key, but it would not change if the account's public key changes.

sunnya97 commented:
I've added the pruning details to the ADR requested by @robert-zaremba and @alexanderbez 

I've also added to the ADR the addition of option 3 from @robert-zaremba's list of proposals on how to support verifiable off-chain signatures, as its the one I find the simplest for clients to implement.

> Update this ADR (ADR-34) to store the past public keys. If someone will rekey, we will have a stack of the previous keys. Dapp can query that stack to search for a public key.

robert-zaremba commented:
> This could be the same public key as the account's public key, but it would not change if the account's public key changes.

I would coin a term: **Special Purpose Account**  :)

sunnya97 commented:
> > This could be the same public key as the account's public key, but it would not change if the account's public key changes.
> 
> I would coin a term: **Special Purpose Account** :)

I was thinking it wouldn't even an Account.  The struct for a Payment channel would look like:

```go

struct PayChan {
    user1Addr        sdk.Address
    user1PubKey   crypto.PubKey

    user2Addr       sdk.Address
    user2Pubkey   crypto.PubKey

    amount
    lastState
    ...
}

```

alexanderbez reviewed (commented):
@sunnya97 let's remove the `EnableChangePubKey` section. Also, what arguments do you have for this feature that CosmWasm would not address.

I think we can keep and merge this ADR as proposed but I think we need more discussion around it, especially if CosmWasm already buys us this functionality.

sunnya97 commented:
@alexanderbez 

While you could use CosmWasm to do rotating keys, there are many SDK-based chains that will probably not want to integrate CosmWasm functionality, due to variety of reasons.  And it will likely be a much while longer before CosmWasm is included on major chains like the Cosmos Hub.  This is a much simpler feature that can be integrated very quickly

> We could instead devise a way to provide a means of one-time "migrating".

If we allow for one time migration, while keeping the same address, we've already accepted the primary negative of this feature (breaking the relationship between pubkey and address).  At that point, what's the point of preventing further rekeying?

robert-zaremba reviewed (approved):
Thanks for updating the description and discussing the outcomes.

alexanderbez commented:
Why just create your own module @sunnya97 and publish it to Atlas? Developers/chains that need such functionality w/o CosmWasm can simple use your module

sunnya97 commented:
@alexanderbez well that's why I proposed having a param that can toggle it on and off, to effectively do that without needing to maintain two different modules
